### ISSUE TITLE
Error Android 12+: Notifications don't open app | Indirect notification activity start (trampoline) from app.package.name blocked

### ISSUE THREAD

**Issue by ollm** [dpa99c/cordova-plugin-firebasex#763]
# Bug report

**CHECKLIST**
* [x] I have read the [issue reporting guidelines](https://github.com/dpa99c/cordova-plugin-firebasex#reporting-issues)
* [x] I confirm this is a suspected bug or issue that will affect other users
* [x] I have reproduced the issue using the [example project](https://github.com/dpa99c/cordova-plugin-firebasex-test) or provided the necessary information to reproduce the issue.
* [x] I have read [the documentation](https://github.com/dpa99c/cordova-plugin-firebasex) thoroughly and it does not help solve my issue.
* [x] I have checked that no similar issues (open or closed) already exist.

**Current behavior:**

Clicking on the notification does not open the app and the following error appears in the logcat

`Indirect notification activity start (trampoline) from app.package.name blocked`

**Expected behavior:**

Clicking on the notification should open the app

**Steps to reproduce:**

Try to open app with target SDK 31 or above from notification on Android 12 or above usin this example payload.

``` json
{
  "data": {
    "notification_body": "Notification with <b>Android</b> settings",
    "notification_title": "Android Notification",
    "notification_foreground": "true",
    "notification_android_body_html": "true",
    "notification_android_id": "a",
    "notification_android_image": "https://user-images.githubusercontent.com/18549191/192135623-14f17f1e-4856-4fc7-a388-d5f380b48707.png",
    "notification_android_image_type": "big_picture",
    "notification_android_channel_id": "fcm_default_channel",
    "key_1": "Data for key one"
  }
}
```

**Console output**
<details>
<summary>console output</summary>

```

2022-09-22 15:39:19.084   561-561   Compatibil...geReporter pid-561                              D  Compat change id reported: 175319604; UID 1000; state: ENABLED
2022-09-22 15:39:19.085   561-1556  ActivityTaskManager     pid-561                              I  START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x14000000 pkg=app.package.name cmp=app.package.name/.MainActivity (has extras)} from uid 10150
2022-09-22 15:39:19.085   864-1109  PeopleSpaceWidgetMgr    pid-864                              D  Sbn doesn't contain valid PeopleTileKey: null/0/app.package.name
2022-09-22 15:39:19.088   561-1556  Compatibil...geReporter pid-561                              D  Compat change id reported: 167676448; UID 10150; state: ENABLED
2022-09-22 15:39:19.088   561-1556  NotificationService     pid-561                              E  Indirect notification activity start (trampoline) from app.package.name blocked
2022-09-22 15:39:19.088   561-1556  ActivityTaskManager     pid-561                              W  Background activity start [callingPackage: app.package.name; callingUid: 10150; appSwitchAllowed: true; isCallingUidForeground: false; callingUidHasAnyVisibleWindow: false; callingUidProcState: RECEIVER; isCallingUidPersistentSystemProcess: false; realCallingUid: 10150; isRealCallingUidForeground: false; realCallingUidHasAnyVisibleWindow: false; realCallingUidProcState: RECEIVER; isRealCallingUidPersistentSystemProcess: false; originatingPendingIntent: null; allowBackgroundActivityStart: false; intent: Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x14000000 pkg=app.package.name cmp=app.package.name/.MainActivity (has extras) }; callerApp: ProcessRecord{dc52746 12283:app.package.name/u0a150}; inVisibleTask: false]
2022-09-22 15:39:19.095  1143-1601  EGL_emulation           pid-1143                             D  app_time_stats: avg=82.47ms min=0.84ms max=1816.03ms count=28
2022-09-22 15:39:19.107   561-1556  ActivityTaskManager     pid-561                              E  Abort background activity starts from 10150
2022-09-22 15:39:19.518   561-982   UserManagerService      pid-561                              V  dumpPackageWhitelistProblems(): using mode ENFORCE|IMPLICIT_WHITELIST|IMPLICIT_WHITELIST_SYSTEM
2022-09-22 15:39:20.082   864-864   ActivityLaunchAnimator  pid-864                              D  Remote animation timed out
2022-09-22 15:39:20.086   864-864   ShadeControllerImpl     pid-864                              V  NotificationShadeWindow: com.android.systemui.statusbar.phone.NotificationShadeWindowView{3083eaf V.E...... ......ID 0,0-1080,2280} canPanelBeCollapsed(): true
2022-09-22 15:39:20.100   864-1103  EGL_emulation           pid-864                              D  app_time_stats: avg=550.81ms min=14.89ms max=1086.72ms count=2
2022-09-22 15:39:20.101   864-1103  EGL_emulation           pid-864                              D  app_time_stats: avg=204.37ms min=0.88ms max=2953.67ms count=17

```

</details>

**Comment by TiBz0u** [comment]
Hi @ollm,

Regarding this [line](https://github.com/dpa99c/cordova-plugin-firebasex/blob/d6b7d469eb2c8e97c6b1d901876893df27cf3da3/src/android/OnNotificationOpenReceiver.java#L31), this plugin seems to use BroadcastReceiver to start Activity.

It seems blocked since Android 12 (first google result : https://proandroiddev.com/notification-trampoline-restrictions-android12-7d2a8b15bbe2).

Can you try to disable the NOTIFICATION_TRAMPOLINE_BLOCK toggle on developer tool to see if it fix the problem?

Kr.

**Comment by ollm** [comment]
With the current version of my app that option cannot be modified, since I have the target SDK 31. I will try to create a version of my app with the target SDK 30 to test with and without the option activated.

Something that I forgot to mention is that to reproduce the error the app has to have the target SDK 31 or above.

<img src="https://user-images.githubusercontent.com/18549191/192112454-a5547dd7-305d-4ff6-83fe-3f558174d756.png" width="300">

**Comment by TiBz0u** [comment]
Hi @ollm , this option is only included for targetSdk 31+ (https://developer.android.com/about/versions/12/reference/compat-framework-changes). I try to figure out how to enable it.
It seems that those flag can be updated only in certain scenario.
Je

**Comment by ollm** [comment]
Hi, from what I've been able to verify, to be able to enable/disable this option it is necessary to have a target SDK in the app lower than 31, since these options are to test the compatibility framework changes that are activated in target SDK 31+ in lower target SDKs.

I have been able to test my app with the target SDK in 30, and the option can be enabled/disabled, with the disabled option I can open the notifications, but once enabled they do not open and the error in the logcat appears.

<img src="https://user-images.githubusercontent.com/18549191/192135623-14f17f1e-4856-4fc7-a388-d5f380b48707.png" width="300">

**Comment by TiBz0u** [comment]
Hi, thanks for your further analysis. I think you need to implement the `PendingIntent` pattern in the plugin. If you have knowledge, maybe you can try a PR.
If not, I can try to make you a fixed forked but I don't think I ll have enough time to test it properly.
Kr

**Comment by ollm** [comment]
Hi, I have made a solution that works correctly on Android 13 (SDK 33) with the app in target SDK 31, I have also tested it on a lower SDK (Android 11, SDK 30) and it works fine, take a look at the code if you want, soon I will open a PR.

https://github.com/ollm/cordova-plugin-firebasex/commit/89d7f026adcd8b986f534e2e4221606f1c0a94b9

**Comment by dpa99c** [comment]
I'm not seeing this issue when testing with the [example project](https://github.com/dpa99c/cordova-plugin-firebasex-test).
For example: [build with API 32 and running on API 33 (screen capture)](https://user-images.githubusercontent.com/2345062/192351077-8be757a0-2bd4-47d7-8cde-88878809cf87.webm)

Can you try building and running the example project and using its [built-in FCM messaging client](https://github.com/dpa99c/cordova-plugin-firebasex-test#messaging-client) to send the test messages?

**Comment by ollm** [comment]
Hi, I have found two ways to reproduce the issue in the example project.

#### The first

Send a notification when the app is in the foreground, when received, move the app to the background and then open the notification.

#### The second

Use this json payload in the notification, the error will always occur, even if the app is in the background, I use a payload similar to this one since I show html and images in my notifications.

 `notification_android.json`
``` json
{
  "data": {
    "notification_body": "Notification with <b>Android</b> settings",
    "notification_title": "Android Notification",
    "notification_foreground": "true",
    "notification_android_body_html": "true",
    "notification_android_id": "a",
    "notification_android_image": "https://user-images.githubusercontent.com/18549191/192135623-14f17f1e-4856-4fc7-a388-d5f380b48707.png",
    "notification_android_image_type": "big_picture",
    "key_1": "Data for key one"
  }
}
```

I understand that the error only occurs when `cordova-plugin-firebasex` handles the notification and not when the `Firebase SDK` handles it

**Comment by dpa99c** [comment]
@ollm Thanks for the info. So the issue is with foreground notifications which are locally-generated by this plugin, as opposed to background/system notifications which are handled by FCM.
Based on your code above, I'll have a look at adding a fix when I get some time this week, unless a PR fixing the issue is submitted in the meantime.

**Comment by ollm** [comment]
Hi, I sent this pull request which in my case fixed the problem, although I haven't tested it in the example project, if you could review the code, thanks.

https://github.com/dpa99c/cordova-plugin-firebasex/pull/765

**Comment by TiBz0u** [comment]
@ollm How the second example can break? If background, by reading your explanation, it will be managed by the FCM.

Does the issue be renamed? If foreground only, you don't have to open the application?

**Comment by ollm** [comment]
@TiBz0u The error also happens in the background if you use the example JSON that I have put, all my notifications use a similar JSON, and in my app the error always happens when the app is in the background.

**Comment by TiBz0u** [comment]
@dpa99c I think I get the reason of this bug (can't open app).

He use data messaging with foreground key set to true. In this way, the broadcastReceiver is use to open the application. However, due to notification trempoline, it's not working.

Kr

**Comment by dpa99c** [comment]
Should be fixed in `cordova-plugin-firebasex@15.0.0` which has just been released to npm
